File: styx-js/src/ledger/chain-validator.js
// styx-js/src/ledger/chain-validator.js
// Chain integrity validation

import { ChainValidationError, ChainErrorType } from './event.js';
import { bytesToHex, utf8Encode } from '../utils.js';
import { StyxPublicKey } from '../crypto/identity.js';

/**
 * Validates the integrity of the ledger chain.
 */
export class ChainValidator {
  /**
   * @param {import('../crypto/hasher.js').Hasher} hasher
   * @param {import('../crypto/signer.js').Verifier} verifier
   */
  constructor(hasher, verifier) {
    this._hasher = hasher;
    this._verifier = verifier;
  }

  /**
   * Validate every event in sequence.
   * @param {import('./event.js').LedgerEvent[]} events - Events ordered by HLC
   * @returns {Promise<ChainValidationError|null>} null if valid
   */
  async validateFullChain(events) {
    if (events.length === 0) return null;

    // Validate genesis
    const genesis = events[0];
    if (genesis.previousHash !== null) {
      return new ChainValidationError(
        genesis.eventId,
        ChainErrorType.GENESIS_VIOLATION,
        'Genesis event must have null previousHash'
      );
    }

    for (let i = 0; i < events.length; i++) {
      const event = events[i];
      const previous = i > 0 ? events[i - 1] : null;
      const pubkey = StyxPublicKey.fromHex(event.senderPubkey);

      const error = await this.validateEvent(event, previous, pubkey);
      if (error) return error;
    }

    return null;
  }

  /**
   * Validate a single event against its predecessor.
   */
  async validateEvent(event, previousEvent, senderPublicKey) {
    // Check hash linkage
    if (previousEvent && event.previousHash !== previousEvent.eventHash) {
      return new ChainValidationError(
        event.eventId,
        ChainErrorType.PREVIOUS_HASH_MISSING,
        `previousHash ${event.previousHash} does not match preceding event hash ${previousEvent?.eventHash}`
      );
    }

    // Verify hash integrity
    const hashValid = await this.verifyEventHash(
      event,
      event.previousHash
    );
    if (!hashValid) {
      return new ChainValidationError(
        event.eventId,
        ChainErrorType.HASH_MISMATCH,
        'Computed hash differs from stored hash'
      );
    }

    // Verify signature
    const sigValid = await this.verifyEventSignature(event, senderPublicKey);
    if (!sigValid) {
      return new ChainValidationError(
        event.eventId,
        ChainErrorType.SIGNATURE_INVALID,
        'Ed25519 signature verification failed'
      );
    }

    // Check HLC monotonicity
    if (previousEvent && event.hlc.compareTo(previousEvent.hlc) <= 0) {
      return new ChainValidationError(
        event.eventId,
        ChainErrorType.HLC_VIOLATION,
        'HLC is not monotonically increasing'
      );
    }

    return null;
  }

  /**
   * Verify event hash matches computed hash
   */
  async verifyEventHash(event, previousHash) {
    const segments = [];
    if (previousHash) segments.push(utf8Encode(previousHash));
    segments.push(utf8Encode(event.eventType));
    if (event.payload) segments.push(event.payload);
    segments.push(event.hlc.toBytes());

    const computed = bytesToHex(this._hasher.compositeHash(segments));
    return computed === event.eventHash;
  }

  /**
   * Verify Ed25519 signature on the event
   */
  async verifyEventSignature(event, publicKey) {
    const hashBytes = this._hasher.hash(utf8Encode(event.eventHash));
    // The signature is over the event hash bytes
    const segments = [];
    if (event.previousHash) segments.push(utf8Encode(event.previousHash));
    segments.push(utf8Encode(event.eventType));
    if (event.payload) segments.push(event.payload);
    segments.push(event.hlc.toBytes());
    const hashBytesComputed = this._hasher.compositeHash(segments);

    return this._verifier.verify(hashBytesComputed, event.signature, publicKey);
  }
}

File: packages/ledger_engine/lib/src/chain_validator.dart
import 'package:meta/meta.dart';
import 'package:styx_crypto_core/styx_crypto_core.dart';
import 'package:styx_ledger_engine/src/event_factory.dart';
import 'package:styx_ledger_engine/src/ledger_event.dart';

/// Type of chain validation error.
enum ChainErrorType {
  /// The computed hash does not match the stored hash.
  hashMismatch,

  /// The Ed25519 signature is invalid.
  signatureInvalid,

  /// The previousHash field does not match the preceding event's hash.
  previousHashMissing,

  /// The HLC is not monotonically increasing.
  hlcViolation,

  /// The first event is not a valid genesis.
  genesisViolation,
}

/// Describes a chain validation error.
@immutable
class ChainValidationError {
  /// Creates a [ChainValidationError].
  const ChainValidationError({
    required this.eventId,
    required this.errorType,
    required this.message,
  });

  /// The event where the error was found.
  final String eventId;

  /// The type of error.
  final ChainErrorType errorType;

  /// Human-readable description.
  final String message;

  @override
  String toString() => 'ChainValidationError($errorType on $eventId)';
}

/// Validates the integrity of the ledger chain.
class ChainValidator {
  /// Creates a [ChainValidator].
  ChainValidator({
    required Hasher hasher,
    required Verifier verifier,
  }) : _eventFactory = EventFactory(signer: Signer(), hasher: hasher),
       _verifier = verifier;

  final EventFactory _eventFactory;
  final Verifier _verifier;

  /// Validates the full chain from start to end.
  ///
  /// Returns the first error found, or `null` if the chain is valid.
  Future<ChainValidationError?> validateFullChain(
    List<LedgerEvent> events,
  ) async {
    if (events.isEmpty) return null;

    // First event must be genesis (no previousHash).
    final genesis = events.first;
    if (genesis.previousHash != null) {
      return ChainValidationError(
        eventId: genesis.eventId,
        errorType: ChainErrorType.genesisViolation,
        message: 'First event must have null previousHash',
      );
    }

    for (var i = 0; i < events.length; i++) {
      final event = events[i];
      final previous = i > 0 ? events[i - 1] : null;
      final senderKey = StyxPublicKey.fromHex(event.senderPubkey);

      final error = await validateEvent(
        event: event,
        previousEvent: previous,
        senderPublicKey: senderKey,
      );
      if (error != null) return error;
    }

    return null;
  }

  /// Validates a single event against its predecessor.
  Future<ChainValidationError?> validateEvent({
    required LedgerEvent event,
    required LedgerEvent? previousEvent,
    required StyxPublicKey senderPublicKey,
  }) async {
    // Check previousHash linkage.
    if (previousEvent != null) {
      if (event.previousHash != previousEvent.eventHash) {
        return ChainValidationError(
          eventId: event.eventId,
          errorType: ChainErrorType.previousHashMissing,
          message: 'previousHash does not match preceding event hash',
        );
      }
    }

    // Verify hash.
    if (!await verifyEventHash(event, event.previousHash)) {
      return ChainValidationError(
        eventId: event.eventId,
        errorType: ChainErrorType.hashMismatch,
        message: 'Computed hash does not match eventHash',
      );
    }

    // Verify signature.
    if (!await verifyEventSignature(event, senderPublicKey)) {
      return ChainValidationError(
        eventId: event.eventId,
        errorType: ChainErrorType.signatureInvalid,
        message: 'Ed25519 signature is invalid',
      );
    }

    // Check HLC monotonicity.
    if (previousEvent != null) {
      if (event.hlc.compareTo(previousEvent.hlc) <= 0) {
        return ChainValidationError(
          eventId: event.eventId,
          errorType: ChainErrorType.hlcViolation,
          message: 'HLC is not monotonically increasing',
        );
      }
    }

    return null;
  }

  /// Verifies that the event's hash matches the computed hash.
  Future<bool> verifyEventHash(
    LedgerEvent event,
    String? previousHash,
  ) async {
    final computed = _eventFactory.computeHashBytes(
      previousHash: previousHash,
      eventType: event.eventType,
      payload: event.payload,
      hlcBytes: event.hlc.toBytes(),
    );
    final computedHex = computed
        .map((b) => b.toRadixString(16).padLeft(2, '0'))
        .join();
    return computedHex == event.eventHash;
  }

  /// Verifies the Ed25519 signature on the event.
  Future<bool> verifyEventSignature(
    LedgerEvent event,
    StyxPublicKey publicKey,
  ) async {
    final hashBytes = _eventFactory.computeHashBytes(
      previousHash: event.previousHash,
      eventType: event.eventType,
      payload: event.payload,
      hlcBytes: event.hlc.toBytes(),
    );
    return _verifier.verify(
      payload: hashBytes,
      signatureBytes: event.signature,
      publicKey: publicKey,
    );
  }
}
